File: core/core-service/src/main/scala/com/dreameddeath/core/service/swagger/CouchbaseDocumentModelConverter.scala
/*
 * Copyright Christophe Jeunesse
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.dreameddeath.core.service.swagger

import com.dreameddeath.core.model.document.{CouchbaseDocument, CouchbaseDocumentElement}
import com.dreameddeath.core.tools.annotation.processor.reflection.{AbstractClassInfo, ClassInfo}
import com.dreameddeath.core.util.{CouchbaseDocumentFieldReflection, CouchbaseDocumentStructureReflection}
import com.wordnik.swagger.converter.{BaseConverter, ModelConverter}
import com.wordnik.swagger.core.SwaggerSpec
import com.wordnik.swagger.model._

import scala.Predef._
import scala.collection.JavaConverters._
import scala.collection.immutable.Map
import scala.collection.mutable
import scala.collection.mutable.{HashMap, LinkedHashMap}

class CouchbaseDocumentModelConverter extends ModelConverter with BaseConverter {
    val m = new HashMap[String, String]

    def isComplex(typeName: String): Boolean = {
        !SwaggerSpec.baseTypes.contains(typeName.toLowerCase)
    }

    override def read(cls: Class[_], typeMap: Map[String, String]): Option[Model]={
        var model: Option[Model] = None
        if(classOf[CouchbaseDocument].isAssignableFrom(cls) || classOf[CouchbaseDocumentElement].isAssignableFrom(cls)){
            implicit val docStructReflection:CouchbaseDocumentStructureReflection = CouchbaseDocumentStructureReflection.getReflectionFromClassInfo(AbstractClassInfo.getClassInfo(cls).asInstanceOf[ClassInfo])
            //implicit val docReflection:CouchbaseDocumentReflection  = CouchbaseDocumentReflection.getReflectionFromClass(cls.asInstanceOf[Class[CouchbaseDocument]])
            val sortedProperties = new LinkedHashMap[String, ModelProperty]()
            for(prop<-docStructReflection.getFields.asScala){
                sortedProperties+=prop.getName -> buildModel(prop,typeMap)
            }
            model = Some(Model(
                getId(docStructReflection),
                docStructReflection.getSimpleName(),
                docStructReflection.getName(),
                sortedProperties,
                Some("")
            ))
        }
        model
    }

    def getId(docReflection:CouchbaseDocumentStructureReflection): String ={
        if(docReflection.getId!=null){
            docReflection.getId
        }
        else{
            docReflection.getSimpleName
        }
    }

    def buildModel(field:CouchbaseDocumentFieldReflection, typeMap: Map[String, String]): ModelProperty={
        val fieldType:String=typeMap.getOrElse(field.getEffectiveTypeClass.getSimpleName.toLowerCase, field.getEffectiveTypeClass.getSimpleName)
        var allowableValues:AllowableValues = AnyAllowableValues
        val isRequired:Boolean = true
        if(fieldType.getClass.isEnum){
            var values:mutable.MutableList[String]=mutable.MutableList.empty[String]
            for(e<-field.getClass.getEnumConstants){
                values+=e.toString //TODO manage more complex json mapping
            }
            allowableValues = AllowableListValues(values.toList)
        }
        var items:Option[ModelRef]=None
        if(field.isCollection || field.isMap){
            val complexType = field.getEffectiveTypeClass.getClass
            val itemType = if(field.isCollection)
                    field.getCollectionElementClass.getSimpleName
            else
                field.getMapValueClass.getSimpleName;

            val itemFullType = if(field.isCollection)
                field.getCollectionElementClass.getName
            else
                field.getMapValueClass.getName;


            if(isComplex(itemType)){
                items = Some(ModelRef(
                    null,Some(itemType),Some(itemFullType)
                ))
            }
            else{
                items = Some(ModelRef(
                    itemType,None,Some(itemFullType)
                ))
            }

        }
        ModelProperty(
            fieldType,
            field.getEffectiveTypeClass.getName,//TODO build better with generics
            0, //TODO manage annotations for position
            isRequired,//TODO manage Required
            Some(""),//Empty Description for now
            allowableValues,
            items
        )
    }


}

File: core/core-service/src/main/java/com/dreameddeath/core/service/swagger/CouchbaseDocumentModelConverter.java
/*
 * Copyright Christophe Jeunesse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.dreameddeath.core.service.swagger;

import com.dreameddeath.compile.tools.annotation.processor.reflection.ParameterizedTypeInfo;
import com.dreameddeath.core.json.ObjectMapperFactory;
import com.dreameddeath.core.model.util.CouchbaseDocumentFieldReflection;
import com.dreameddeath.core.model.util.CouchbaseDocumentStructureReflection;
import com.dreameddeath.core.service.utils.ServiceObjectMapperConfigurator;
import com.fasterxml.jackson.databind.JavaType;
import io.swagger.converter.ModelConverter;
import io.swagger.converter.ModelConverterContext;
import io.swagger.jackson.AbstractModelConverter;
import io.swagger.jackson.TypeNameResolver;
import io.swagger.models.Model;
import io.swagger.models.ModelImpl;
import io.swagger.models.properties.ArrayProperty;
import io.swagger.models.properties.MapProperty;
import io.swagger.models.properties.Property;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import java.util.Iterator;

/**
 * Created by Christophe Jeunesse on 24/04/2015.
 */
public class CouchbaseDocumentModelConverter extends AbstractModelConverter {
    private static Logger LOG = LoggerFactory.getLogger(CouchbaseDocumentModelConverter.class);

    public CouchbaseDocumentModelConverter() {
        super(ObjectMapperFactory.BASE_INSTANCE.getMapper(ServiceObjectMapperConfigurator.SERVICE_MAPPER_CONFIGURATOR));
    }


    @Override
    public Model resolve(Type type, ModelConverterContext modelConverterContext, Iterator<ModelConverter> iterator) {
        Class rawClass = null;
        if(type instanceof JavaType){
            rawClass=((JavaType)type).getRawClass();
        }
        else if(type instanceof Class){
            rawClass = (Class)type;
        }
        if((rawClass!=null) && CouchbaseDocumentStructureReflection.isReflexible(rawClass)){
            CouchbaseDocumentStructureReflection eltReflexion = CouchbaseDocumentStructureReflection.getReflectionFromClassInfo(rawClass);
            ModelImpl model = new ModelImpl()
                    .name(TypeNameResolver.std.nameForType(_mapper.constructType(type)));
            for(CouchbaseDocumentFieldReflection field:eltReflexion.getFields()){
                String propName = field.getName();
                Property prop;
                if(field.isCollection()) {
                    prop = new ArrayProperty()
                            .items(resolveProperty(field.getCollectionElementTypeInfo(),field.getField().getAnnotations(),modelConverterContext,iterator));
                }
                else if(field.isMap()){
                    prop = new MapProperty()
                            .additionalProperties(resolveProperty(field.getMapValueTypeInfo(),field.getField().getAnnotations(),modelConverterContext,iterator));
                }
                else{
                    prop = resolveProperty(field.getEffectiveTypeInfo(),field.getField().getAnnotations(),modelConverterContext,iterator);
                }
                prop.setName(propName);
                model.addProperty(propName,prop);
                //model.addRequired(propName);
                //prop.setRequired(field.getField().getAnnotation(NotNull));
            }
            //Todo add description

            LOG.debug("founding class {}", type);

            return model;
        }
        return super.resolve(type, modelConverterContext, iterator);
    }

    protected Property resolveProperty(ParameterizedTypeInfo info,Annotation[] annotations,ModelConverterContext modelConverterContext, Iterator<ModelConverter> iterator){
        return modelConverterContext.resolveProperty(info.getType(),annotations);
    }
}
